wireless_modules/wireless_module.py

```diff
@@ -0,0 +1,100 @@
+"""
+A class structure to read and collate data from different sensors into a dictionary and send through MQTT
+"""
+import ujson
+import utime
+from mqtt_client import Client
+
+
+class WirelessModule:
+    def __init__(self, client_id, broker):
+        """
+        Initialises the wireless module
+        :param client_id: A string representing the mqtt client id
+        :param broker: A string representing the IP address or web domain of the mqtt broker
+        """
+        self.sensors = None
+        self.start_publish = False
+        self.sub_topics = []
+        self.pub_topics = []
+
+        self.mqtt = Client(client_id, broker)
+
+    def set_sub_topics(self, start_topic, stop_topic):
+        """
+        Set the MQTT topics to subscribe to.
+        :param start_topic: A byte literal of the topic to wait for the start message
+        :param stop_topic: A byte literal of the topic to check for the stop message
+        """
+        if not isinstance(start_topic, bytes) or not isinstance(stop_topic, bytes):
+            raise Exception("Ensure the subscribe topics specified are byte literals")
+        self.sub_topics.append(start_topic)
+        self.sub_topics.append(stop_topic)
+
+    def set_pub_topics(self, data_topic, battery_topic, low_battery_topic):
+        """
+        Set the MQTT Topics to publish to.
+        :param data_topic: A byte literal representing the topic to publish sensor data to
+        :param battery_topic: A byte literal representing the topic to publish battery percentage level to
+        :param low_battery_topic: A byte literal representing the topic to publish to when battery is low
+        """
+        self.pub_topics.append(data_topic)
+        self.pub_topics.append(battery_topic)
+        self.pub_topics.append(low_battery_topic)
+
+    def add_sensors(self, sensor_arr):
+        """
+        Store instances of sensor class
+        :param sensor_arr: An array of sensor class instances.
+        :pre-requisites: Each class must have a .read() method to read data through
+        """
+        self.sensors = sensor_arr
+
+    def _read_sensors(self):
+        """
+        Reads sensor data from each sensor object stored within this class instance
+        :return: A dictionary of all the sensor types and they're corresponding sensor reading/s
+        :pre-requisite: The read() method for each sensor must return a dictionary
+        """
+        readings = {"sensors": []}
+
+        for sensor in self.sensors:
+            sensor_data = sensor.read()
+            for data in sensor_data:
+                readings["sensors"].append(data)
```

Comment:
Forgive me if my understanding is incorrect, because I don't have an ESP32 to test this code.  But couldn't you just append `sensor_data` into the `"sensors"` array without needing to loop through `sensor_data`?

Reply:
This is done because a sensor may return multiple bits of data, e.g. the DHT22 returns both temperature and humidity. So, we need to loop through any returned data and add them separately.